### PR TITLE
Issue #2316 - support deprecated 'helm.repositories' config

### PR DESCRIPTION
**Pull request by alexmt**

### CODE REVIEW

File: test/e2e/helm_test.go [hunk 1]
```diff
@@ -178,3 +184,46 @@ func TestKubeVersion(t *testing.T) {
...
+		ctx.And(func() {
+			errors.FailOnErr(fixture.Run("", "kubectl", "create", "secret", "generic", "helm-repo",
+				"-n", fixture.ArgoCDNamespace,
+				fmt.Sprintf("--from-file=certSecret=%s", repos.CertPath),
+				fmt.Sprintf("--from-file=keySecret=%s", repos.CertKeyPath),
+				fmt.Sprintf("--from-literal=username=%s", GitUsername),
+				fmt.Sprintf("--from-literal=password=%s", GitPassword),
+			))
+			errors.FailOnErr(fixture.KubeClientset.CoreV1().Secrets(fixture.ArgoCDNamespace).Patch(
+				"helm-repo", types.MergePatchType, []byte(`{"metadata": { "labels": {"e2e.argoproj.io": "true"} }}`)))
```
Comment (alexec) [inline test/e2e/helm_test.go]: nice

File: util/db/db.go [hunk 2]
```diff
@@ -41,6 +42,9 @@ type ArgoDB interface {
 	CreateRepoCertificate(ctx context.Context, certificate *appv1.RepositoryCertificateList, upsert bool) (*appv1.RepositoryCertificateList, error)
 	// CreateRepoCertificate creates a new certificate entry
 	RemoveRepoCertificates(ctx context.Context, selector *CertificateListSelector) (*appv1.RepositoryCertificateList, error)
+
+	// ListHelmRepositories lists repositories
+	ListHelmRepositories(ctx context.Context) ([]*appv1.Repository, error)
```
Comment (alexec) [inline util/db/db.go]: // DERECATED

File: util/db/helmrepository.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	repoInfo := helmRepositories[index]
+
+	repo := &v1alpha1.Repository{
+		Repo:                  repoInfo.URL,
+		Type:                  "helm",
+		Name:                  repoInfo.Name,
+		InsecureIgnoreHostKey: false,
+		Insecure:              false,
+		EnableLFS:             false,
```
Comment (alexec) [inline util/db/helmrepository.go]: delete?

File: util/db/helmrepository.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	err := db.unmarshalFromSecretsStr(map[*string]*v1.SecretKeySelector{
+		&repo.Username:          repoInfo.UsernameSecret,
+		&repo.Password:          repoInfo.PasswordSecret,
+		&repo.TLSClientCertData: repoInfo.CertSecret,
+		&repo.TLSClientCertKey:  repoInfo.KeySecret,
```
Comment (alexec) [inline util/db/helmrepository.go]: what about CA?

File: util/settings/settings.go [hunk 5]
```diff
@@ -94,6 +94,15 @@ type OIDCConfig struct {
 	RequestedIDTokenClaims map[string]*oidc.Claim `json:"requestedIDTokenClaims,omitempty"`
 }
 
+type HelmRepoCredentials struct {
```
Comment (alexec) [inline util/settings/settings.go]: // deprecated?

File: util/settings/settings.go [hunk 6]
```diff
@@ -316,6 +327,22 @@ func (mgr *SettingsManager) GetKustomizeBuildOptions() (string, error) {
 	return argoCDCM.Data[kustomizeBuildOptions], nil
 }
 
+func (mgr *SettingsManager) GetHelmRepositories() ([]HelmRepoCredentials, error) {
```
Comment (alexec) [inline util/settings/settings.go]: // deprecated?

File: util/db/helmrepository.go (hunk 3 above)
Comment (alexmt) [reply util/db/helmrepository.go]: done

File: util/settings/settings.go (hunk 5 above)
Comment (alexmt) [reply util/settings/settings.go]: done

File: util/settings/settings.go (hunk 6 above)
Comment (alexmt) [reply util/settings/settings.go]: done

File: util/db/db.go (hunk 2 above)
Comment (alexmt) [reply util/db/db.go]: This one is not really deprecated. Even after we remove legacy helm repos it still makes sense to use this method to get only repos with helm type. In one of our Argo CD instances, we have ~350 git repos and 0 helm repos. So it does not makes sense to send 350 repos to Repo Server, instead it is better to filter out non-helm repos on the client side.

File: util/db/helmrepository.go (hunk 4 above)
Comment (alexmt) [reply util/db/helmrepository.go]: Nice catch. It is impossible to support/migrate configured CA with the current design. 

The good thing that CA support does not really work in helm itself: https://github.com/helm/helm/issues/3384 . So unlikely anyone managed to configure custom CA in Argo CD and we don't have to worry about this case.

File: util/settings/settings.go (hunk 5 above)
Comment (alexec) [reply util/settings/settings.go]: I think you might be able to extend `Repository` in Go

File: util/db/helmrepository.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+	result := make([]*v1alpha1.Repository, len(helmRepositories))
+	for i, helmRepoInfo := range helmRepositories {
+		repo, err := db.getHelmRepo(helmRepoInfo.URL, helmRepositories)
+		if err != nil {
+			return nil, err
+		}
+		result[i] = repo
+	}
+	repos, err := db.ListRepositories(ctx)
```
Comment (alexec) [inline util/db/helmrepository.go]: good stuff

File: util/db/helmrepository.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	if err != nil {
+		return nil, err
+	}
+	for i := range repos {
```
Comment (alexec) [inline util/db/helmrepository.go]: do we want to add a `Filter` func to `Repositories`?

**Review by alexec — APPROVED** [review]
Some minor comments.

File: util/db/helmrepository.go (hunk 8 above)
Comment (alexmt) [reply util/db/helmrepository.go]: I like this idea. added

**Pull request merged**